File: statsig-ffi/bindings/ruby/statsig.rb
require_relative 'statsig_ffi'

class User
  def initialize(name, email)
    @user_ptr = StatsigFFI.create_user(name, email)
  end

  def self.create(name, email)
    new(name, email)
  end

  def destroy
    StatsigFFI.destroy_user(@user_ptr) if @user_ptr
  end

  def self.destroy(user)
    user.destroy
  end
end

class Statsig
  def initialize(user, sdk_key)
    @statsig_ptr = StatsigFFI.create_statsig_for_user(user.instance_variable_get(:@user_ptr), sdk_key)
  end

  def self.for_user(user, sdk_key)
    new(user, sdk_key)
  end

  def destroy
    StatsigFFI.destroy_statsig(@statsig_ptr) if @statsig_ptr
  end

  def self.destroy(statsig)
    statsig.destroy
  end

  def check_gate(gate_name)
    StatsigFFI.check_gate(@statsig_ptr, gate_name)
  end

  def get_client_init_res
    StatsigFFI.statsig_get_client_init_response(@statsig_ptr)
  end
end

File: statsig-ffi/bindings/python/statsig.py
from _statsig_ffi import ffi, lib

class User:
    def __init__(self, name: str, email: str):
        self._user = lib.create_user(name.encode('utf-8'), email.encode('utf-8'))

    def __del__(self):
        if hasattr(self, '_user'):
            lib.destroy_user(self._user)


class Statsig:
    def __init__(self, sdk_key: str):
        self._statsig = lib.initialize_statsig(sdk_key.encode('utf-8'))

    def __del__(self):
        if hasattr(self, '_statsig'):
            lib.destroy_statsig(self._statsig)

    def check_gate(self, user: User, gate_name: str) -> bool:
        return lib.check_gate(self._statsig, user._user, gate_name.encode('utf-8'))

    def get_client_init_response(self, user: User) -> str:
        res = lib.statsig_get_client_init_response(self._statsig, user._user)
        return ffi.string(res).decode('utf-8')
